akka-cluster-sharding/src/test/scala/akka/cluster/sharding/LeastShardAllocationStrategySpec.scala

```diff
@@ -162,27 +179,47 @@ class LeastShardAllocationStrategySpec extends AkkaSpec("akka.actor.provider=clu
     final class DummyActorRef(val path: ActorPath) extends MinimalActorRef {
       override def provider: ActorRefProvider = ???
     }
-    "filter out old version nodes when during rolling update" in {
-      val selfMember =
+
+    "prefer least shards, latest version, non downed, leaving or exiting nodes" in {
+      // old version, up
+      val oldMember =
         Member(UniqueAddress(Address("akka", "myapp", "127.0.0.1", 252525), 1L), Set.empty, Version("1.0.0"))
-      val memberA =
-        Member(UniqueAddress(Address("akka", "myapp", "127.0.0.2", 252525), 1L), Set.empty, Version("1.0.0"))
-      val memberB =
+          .copy(MemberStatus.Up)
+      // leaving, new version
+      val leavingMember =
+        Member(UniqueAddress(Address("akka", "myapp", "127.0.0.2", 252525), 1L), Set.empty, Version("1.0.1"))
+          .copy(MemberStatus.Leaving)
+      // new version, up
+      val newVersionMember1 =
         Member(UniqueAddress(Address("akka", "myapp", "127.0.0.3", 252525), 1L), Set.empty, Version("1.0.1"))
-
-      val fakeLocalShard = new DummyActorRef(RootActorPath(Address("akka", "myapp")) / "system" / "fake" / "shard")
-      val fakeShardA = new DummyActorRef(RootActorPath(memberA.address) / "system" / "fake" / "shard")
-      val fakeShardB = new DummyActorRef(RootActorPath(memberB.address) / "system" / "fake" / "shard")
-
-      val allocationMap: LeastShardAllocationStrategy.AllocationMap =
-        Map(fakeLocalShard -> Vector.empty, fakeShardA -> Vector.empty, fakeShardB -> Vector.empty)
-      val result = LeastShardAllocationStrategy.onlyShardsOfHighestVersion(
-        allocationMap,
-        selfMember.address,
-        SortedSet(selfMember, memberA, memberB))
+          .copy(MemberStatus.Up)
+      // new version, up
+      val newVersionMember2 =
+        Member(UniqueAddress(Address("akka", "myapp", "127.0.0.4", 252525), 1L), Set.empty, Version("1.0.1"))
+          .copy(MemberStatus.Up)
+
+      val fakeLocalShard = new DummyActorRef(RootActorPath(Address("akka", "myapp")) / "system" / "fake" / "oldapp")
+      val fakeShardA = new DummyActorRef(RootActorPath(leavingMember.address) / "system" / "fake" / "leaving")
+      val fakeShardB = new DummyActorRef(RootActorPath(newVersionMember1.address) / "system" / "fake" / "fewest")
+      val fakeShardC = new DummyActorRef(RootActorPath(newVersionMember2.address) / "system" / "fake" / "oneshard")
+
+      val shardsAndMembers: Seq[(ActorRef, Member, immutable.IndexedSeq[ShardId])] =
+        Seq(
+          (fakeLocalShard, oldMember, Vector.empty),
+          (fakeShardA, leavingMember, Vector.empty),
+          (fakeShardB, newVersionMember1, Vector.empty),
+          (fakeShardC, newVersionMember2, Vector("ShardId1")))
```

Comment:
If we swap these two, than the test will prove that the size of the vector is be used in the sorting as well. 

As it is know, one could argue that only the status and the version are at play.

Reply:
Um, no, both member a and b are up and the same version, so these last two is only about choosing the region with least shards

Reply:
My point was if you remove the sorting on the Vector, B and C will get the same ordering after the sorting. 

We need to have a region with more shards in front of one we less so the test verifies that the impl is effectively sorting by num of shards.

Reply:
Ah, I see, I'll re-order or add another region even to really make sure.